documentation/rfc7700.md

```diff
@@ -0,0 +1,164 @@
+
+[precis]: https://tools.ietf.org/html/rfc7700#section-2
+
+
+### Order of Operations
+
+Names being prepared MUST apply the following rules in the order shown:
+
+1. Preperation using the PRECIS [Nickname profile][precis].
+2. Check for restricted characters.
+
+These steps MUST happen in the order shown, or else the restricted characters check may miss characters that should be legitimately restricted.
+
+If a name does contain a restricted character (whether disallowed by the [Nickname profile](https://tools.ietf.org/html/rfc7700#section-2.2) or this document), it MUST be rejected by the server and MUST NOT be propogated to other clients. This is done through the appropriate numeric for the command which tried to set or use the invalid name such as `ERR_ERRONEUSNICKNAME`, `ERR_NOSUCHCHANNEL`, or whichever numeric is most appropriate.
+
+
+### Comparisons
+
+When comparing user names or channel names to check for equivalency, both strings are casemapped with the above process. After this, they are compared with a standard bit-for-bit comparison.
+
+
+## Restricted Characters
+
+Certain characters mean special things to the protocol. This section details characters that are not allowed in specific name types.
+
+
+### All Names
+
+Nicknames, usernames, hostnames and channel names cannot contain the following characters:
+
+* `(' ', 0x20)` - Separates parameters.
+* `(',', 0x2C)` - Used as a separator.
+* `('*', 0x2A)` - Used in mask matching.
+* `('?', 0x3F)` - Used in mask matching.
+
+In addition, the first character of nicknames, usernames, hostnames and channel names cannot be:
+
+* `(':', 0x3A)` - Separates trailing parameter.
+
+To be clear, these characters will also be repeated below.
+
+#### Protocol Framing
+
+These characters are disallowed either by the protocol itself or because of how they are used in protocol framing. These won't be repeated below, as sane implementations either break the message on or disallow the character in streams:
+
+* `('\0', 0x00)` - Disallowed by the protocol.
+* `('\n', 0x0A)` - Used in message framing.
+* `('\r', 0x0D)` - Used in message framing.
+
+
+### Nicknames
+
+Nicknames cannot contain the following characters:
+
+* `(' ', 0x20)` - Separates parameters.
+* `(',', 0x2C)` - Used as a separator.
+* `('*', 0x2A)` - Used in mask matching.
+* `('?', 0x3F)` - Used in mask matching.
+* `('.', 0x2E)` - Denotes a server name.
+* `('!', 0x21)` - Separates nickname from username.
+* `('@', 0x40)` - Separates username from hostname.
+
+In addition to the above restrictions, the first character of a nickname cannot be:
+
+* `(':', 0x3A)` - Separates trailing parameter.
+* `('-', 0x2D)` - Disallowed.
+* `('0', 0x30)` - Disallowed.
+* `('1', 0x31)` - Disallowed.
+* `('2', 0x32)` - Disallowed.
+* `('3', 0x33)` - Disallowed.
+* `('4', 0x34)` - Disallowed.
+* `('5', 0x35)` - Disallowed.
+* `('6', 0x36)` - Disallowed.
+* `('7', 0x37)` - Disallowed.
+* `('8', 0x38)` - Disallowed.
+* `('9', 0x39)` - Disallowed.
```

Comment:
There's nothing stopping servers from accepting a subset of nicks allowed by this spec (they can send an invalid nick numeric for any nick they don't like) so servers can still disallow digits if they want but they cannot allow more nicks than what this spec allows. Also clients must be prepared to see nicks starting with digits.